File: scala/nominal_typing.scala
object main1 extends App {
  abstract class Animal {
    def sound(): Unit
  }


  class Cat extends Animal {
    def sound(): Unit = {
      println("Meow meow!")
    }
  }


  class Dog {
    def sound(): Unit = {
      println("Woof woof!")
    }
  }


  def makeSound(animal: Animal): Unit = {
    animal.sound()
  }

  makeSound(new Cat())
  makeSound(new Dog())
}

File: python/nominal_typing.py
from abc import ABC, abstractmethod


class Animal(ABC):
    @abstractmethod
    def sound(self):
        ...


class Cat(Animal):
    def sound(self):
        print('Meow meow!')


class Dog:
    def sound(self):
        print('Woof woof!')


def make_sound(animal: Animal):
    animal.sound()


make_sound(Cat())
make_sound(Dog())
